internal/storage/secret.go

```diff
@@ -0,0 +1,237 @@
+	Name   string    `db:"name" json:"name"`     // Name is cardholder name.
+	Number string    `db:"number" json:"number"` // Number is card number.
+	Date   string    `db:"date" json:"date"`     // Date is card expiration date.
+	CVV    string    `db:"cvv" json:"cvv"`       // CVV is CVV (or CVC).
+}
+
+// CreateSecret creates a new secret in DB.
+func (s *PgSQL) CreateSecret(ctx context.Context, secret *Secret) error {
+	return WithVoidTransaction(ctx, s, func(tx pgx.Tx) error {
+		_, ok := api.Kinds[secret.Kind]
+		if !ok {
+			return ErrInvalidKind
+		}
+
+		if err := createSecret(ctx, tx, secret); err != nil {
+			return err
+		}
+
+		return tx.Commit(ctx)
+	})
+}
+
+// DeleteSecret deletes a secret from a DB.
+func (s *PgSQL) DeleteSecret(ctx context.Context, secretID uuid.UUID) error {
+	query := `delete from public.secret where id = $1`
+	_, err := s.Conn.Exec(ctx, query, secretID)
+	return err
+}
+
+// LoadSecretByName loads a secret by name.
+func (s *PgSQL) LoadSecretByName(ctx context.Context, userID uuid.UUID, name string) (*Secret, error) {
+	return WithTransaction(ctx, s, func(tx pgx.Tx) (*Secret, error) {
+		result, err := loadSecretByName(ctx, tx, userID, name)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if err := tx.Commit(ctx); err != nil {
+			return nil, err
+		}
+
+		return result, err
+	})
+}
+
+// LoadSecretByID loads a secret by ID.
+func (s *PgSQL) LoadSecretByID(ctx context.Context, secretID uuid.UUID) (*Secret, error) {
+	return WithTransaction(ctx, s, func(tx pgx.Tx) (*Secret, error) {
+		result, err := loadSecretByID(ctx, tx, secretID)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if err := tx.Commit(ctx); err != nil {
+			return nil, err
+		}
+
+		return result, err
+	})
+}
+
+// LoadSecrets loads all secrets for given user.
+func (s *PgSQL) LoadSecrets(ctx context.Context, userID uuid.UUID) (*[]Secret, error) {
+	var rows []Secret
+
+	query := `
+		select
+			s.*,
+			json_agg_strict(t.text) tags
+		from secret s
+		left join tag t on s.id = t.secret_id
+		where s.user_id = $1
+		group by s.id
+		order by s.name
+	`
+	err := pgxscan.Select(ctx, s.Conn, &rows, query, userID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &rows, nil
+}
+
+// RenameSecret renames secret.
+func (s *PgSQL) RenameSecret(ctx context.Context, secretID uuid.UUID, name string) error {
+	query := `update public.secret set name = $1 where id = $2`
+	_, err := s.Conn.Exec(ctx, query, name, secretID)
+	if err != nil {
+		var pgErr *pgconn.PgError
+		if errors.As(err, &pgErr) && pgErr.Code == "23505" {
+			return ErrDuplicateSecretFound
+		}
+		return err
+	}
+
+	return nil
+}
+
+func loadSecretByName(ctx context.Context, tx pgx.Tx, userID uuid.UUID, name string) (*Secret, error) {
+	var secret Secret
+
+	query := `
+		select
+			s.*,
+			json_agg_strict(t.text) tags
+		from secret s
+		left join tag t on s.id = t.secret_id
+		where s.user_id = $1 and s.name = $2
+		group by s.id
```

Comment:
```suggestion
		group by s.id
		for update
```
Однако, необходимо проверить, будет ли работать `for update` вместе с `group by`. Но без него в конфигурации PostgreSQL по умолчанию запись не будет заблокирована в рамках транзакции.

Reply:
Не, с группировкой `for update` не работает. Да здесь транзакция и не нужна, это место только для получения используется. И в принципе транзакции в этом проекте не сильно нужны, это ведь не OLTP-система. Не будет один юзер одновременно с двух клиентов что-то такое делать, что может законфликтиться